internal/provider/connection/connection.go

```diff
@@ -0,0 +1,393 @@
+					for k := range i.(map[string]interface{}) {
+						if k != strings.ToLower(k) {
+							nonLowerCasedKeys = append(nonLowerCasedKeys, k)
+						}
+					}
+					if len(nonLowerCasedKeys) > 0 {
+						return diag.Errorf("data keys must be lowercase: %s", strings.Join(nonLowerCasedKeys, ","))
+					}
+					return nil
+				},
+				Elem: &schema.Schema{
+					Required: true,
+					Type:     schema.TypeString,
+				},
+			},
+		},
+	}
+}
+
+func resourceClusterConnectionCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	var (
+		c                       = meta.(*scylla.Client)
+		dcName                  = d.Get("datacenter").(string)
+		cidrListVal, cidrListOK = d.GetOk("cidrlist")
+		clusterID               = d.Get("cluster_id").(int)
+		r                       = &model.ClusterConnectionCreateRequest{
+			Name: d.Get("name").(string),
+			Data: schemautils.ConvertMapToConcrete[string](d.Get("data").(map[string]interface{})),
+			Type: d.Get("type").(string),
+		}
+		p *scylla.CloudProvider
+	)
+
+	dcs, err := c.ListDataCenters(ctx, int64(clusterID))
+	if err != nil {
+		return diag.Errorf("error reading clusters: %s", err)
+	}
+
+	for _, dc := range dcs {
+		if strings.EqualFold(dc.Name, dcName) {
+			r.ClusterDCID = dc.ID
+			p = c.Meta.ProviderByID(dc.CloudProviderID)
+			if p == nil {
+				return diag.Errorf("unable to find cloud provider with id=%d", dc.CloudProviderID)
+			}
+			break
+		}
+	}
+
+	if r.ClusterDCID == 0 {
+		return diag.Errorf("unable to find %q datacenter", dcName)
+	}
+
+	if !cidrListOK {
+		return diag.Errorf(`"cidrlist" is required for %q cloud`, p.CloudProvider.Name)
+	}
+
+	if len(cidrListVal.([]any)) == 0 {
+		return diag.Errorf(`"cidrlist" cannot be empty`)
+	}
+
+	r.CIDRList, err = schemautils.ConvertListToConcrete[string](cidrListVal)
+	if err != nil {
+		return diag.Errorf(`"cidrlist" must be a list of strings`)
+	}
+
+	conn, err := c.CreateClusterConnection(ctx, int64(clusterID), r)
+	if err != nil {
+		return diag.Errorf("error creating cluster connection: %s", err)
+	}
+	d.SetId(strconv.FormatInt(conn.ID, 10))
+	err = waitForClusterConnection(ctx, c, int64(clusterID), conn.ID, "ACTIVE")
+	if err != nil {
+		return diag.Errorf(err.Error())
+	}
+	conn, err = c.GetClusterConnection(ctx, int64(clusterID), conn.ID)
+	if err != nil {
+		return diag.Errorf("error reading cluster connection %d: %s", conn.ID, err)
+	}
+	_ = d.Set("external_id", conn.ExternalID)
+	_ = d.Set("status", conn.Status)
+	return nil
+}
+
+var errNotFound = errors.New("not found")
+
+func resourceClusterConnectionRead(ctx context.Context, d *schema.ResourceData, meta interface{}) error {
+	var (
+		c         = meta.(*scylla.Client)
+		connIDStr = d.Id()
+		dc        *model.Datacenter
+	)
+
+	if connIDStr == "" {
+		return nil
+	}
+
+	connectionID, err := strconv.ParseInt(connIDStr, 10, 64)
+	if err != nil {
+		return fmt.Errorf("failed to parse connection id %q: %s", connIDStr, err)
```

Comment:
```suggestion
		return fmt.Errorf("failed to parse connection id %q: %w", connIDStr, err)
```

usually you want to wrap the error, just in case it's needed to examine it later